### PR TITLE
While loops in GTC toolchain

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Description

This PR implements the `while` loop construct in the GTC toolchain and relies on [PR #410](https://github.com/GridTools/gt4py/pull/410) for the frontend implementation.

Blocked by #628.

### CODE REVIEW

**Comment by DropD** [comment]
bors try

File: src/gt4py/backend/python_generator.py [hunk 1]
```diff
@@ -16,6 +16,8 @@
 
 import types
 
+from numpy import set_printoptions
```
Comment (jdahm) [inline src/gt4py/backend/python_generator.py:19]: Probably don't need this.

File: src/gtc/cuir/oir_to_cuir.py [hunk 2]
```diff
@@ -90,7 +90,14 @@ def visit_AssignStmt(self, node: oir.AssignStmt, **kwargs: Any) -> cuir.AssignSt
 
     def visit_MaskStmt(self, node: oir.MaskStmt, **kwargs: Any) -> cuir.MaskStmt:
         return cuir.MaskStmt(
-            mask=self.visit(node.mask, **kwargs), body=self.visit(node.body, **kwargs)
+            mask=self.visit(node.mask, **kwargs),
+            body=self.visit(node.body, **kwargs),
```
Comment (jdahm) [inline src/gtc/cuir/oir_to_cuir.py:94]: Removing the `,` will put this on the same line. Same below.

File: src/gtc/gtir.py [hunk 3]
```diff
@@ -145,6 +145,31 @@ def verify_scalar_condition(cls, cond: Expr) -> Expr:
...
 
+class While(common.While[Stmt, Expr], Stmt):
+    """
+    While loop with a field or scalar expression as condition.
+
+    No special rules apply.
+    """
+
+    @validator("body")
+    def verify_no_accumulated_extents(cls, body: List[Stmt]) -> List[Stmt]:
```
Comment (jdahm) [inline src/gtc/gtir.py:156]: Note: this is a much better check than the one for the `For` check in #487.

File: tests/test_unittest/test_gtc/gtir_utils.py [hunk 4]
```diff
@@ -88,6 +88,14 @@ class Meta:
     false_branch = None
 
 
+class WhileFactory(factory.Factory):
+    class Meta:
+        model = gtir.While
+
+    cond = factory.SubFactory(BinaryOpFactory, dtype=common.DataType.BOOL)
+    body = []
```
Comment (jdahm) [inline tests/test_unittest/test_gtc/gtir_utils.py:96]: Add type hint.

File: src/gtc/gtir_to_oir.py [hunk 5]
```diff
@@ -54,31 +49,44 @@ class Context:
         they attach their result to the Context object.
         """
 
-        decls: List = field(default_factory=list)
-        horizontal_executions: List = field(default_factory=list)
+        class ContextInScope:
```
Comment (jdahm) [inline src/gtc/gtir_to_oir.py:52]: Use the version in the for loop PR instead of this.

File: src/gtc/dace/expansion.py [hunk 6]
```diff
@@ -163,6 +163,19 @@ def visit_MaskStmt(self, node: oir.MaskStmt, **kwargs):
         body_code = [indent + b for b in body_code]
         return "\n".join([mask_str] + body_code)
 
+    def visit_While(self, node: oir.While, **kwargs):
+        body = self.visit(node.body, **kwargs)
+        cond = self.visit(node.cond, is_target=False, **kwargs)
+
+        init = "num_iter = 0"
+        max_iter = 10000
+        cond += f" and (num_iter < {max_iter})"
```
Comment (jdahm) [inline src/gtc/dace/expansion.py:172]: Why is this needed?

File: src/gtc/dace/expansion.py [hunk 7]
```diff
@@ -163,6 +163,19 @@ def visit_MaskStmt(self, node: oir.MaskStmt, **kwargs):
...
+        body.append("num_iter += 1")
```
Comment (jdahm) [inline src/gtc/dace/expansion.py:173]: Curious: why is this needed?

Comment (eddie-c-davis) [reply src/gtc/dace/expansion.py:173]: This is needed because DaCe-orchestrated mode can only run without halo exchanges at this time, and without halo exchanges, the `lagrangian_contributions` in `fv3core` will loop indefinitely.

File: src/gtc/gtir.py [hunk 8]
```diff
@@ -149,6 +149,31 @@ def verify_scalar_condition(cls, cond: Expr) -> Expr:
...
+        offset_reads = set()
+        field_writes = set()
+        for stmt in body:
+            field_writes |= stmt.iter_tree().if_isinstance(FieldAccess).getattr("name").to_set()
```
Comment (jdahm) [inline src/gtc/gtir.py:164]: I believe this gathers all accesses, not just writes.

File: src/gtc/gtir_to_oir.py (hunk 5 above)
Comment (jdahm) [reply src/gtc/gtir_to_oir.py:52]: This is the portion of this PR that still needs a rewrite.

File: src/gtc/dace/expansion.py (hunk 7 above)
Comment (jdahm) [reply src/gtc/dace/expansion.py:173]: Sounds good. Would it be OK with you if we moved this to another branch for the time being, so that we don't mix fv3core and gt4py concerns?

**Comment by jdahm** [comment]
bors try

File: src/gtc/gtir_to_oir.py [hunk 9]
```diff
@@ -18,192 +18,177 @@
...
-    ) -> oir.VerticalLoop:
-        ctx.horizontal_executions.clear()
-        self.visit(node.body, ctx=ctx)
+    # --- Control flow ---
+    def visit_VerticalLoop(self, node: gtir.VerticalLoop, *, ctx: Context) -> oir.VerticalLoop:
+        horiz_execs: List[oir.HorizontalExecution] = []
+        for stmt in node.body:
+            ctx.reset_local_scalars()
+            ret = self.visit(stmt, ctx=ctx)
+            stmts = utils.flatten_list([ret] if isinstance(ret, oir.Stmt) else ret)
```
Comment (jdahm) [inline src/gtc/gtir_to_oir.py:169]: I don't love this "hack", but otherwise I think the lowering is improved by this change.

**Comment by jdahm** [comment]
This PR is ready to go. Could someone from CSCS review it when they get a chance?

**Review by havogt — CHANGES_REQUESTED** [review]
I didn't review everything yet.

Before I go deeper, could you summarize what the limitations are of while loops?
- Should nested loops be allowed? I made an example that fails in the parsing.
- From the validator, I assume we are conservative for safety, and e.g. allow no offsets that could trigger race conditions in blocked execution. That's good. Can you extend the parallel model with a section about while loops?

**Comment by jdahm** [comment]
> ... could you summarize what the limitations are of while loops?

Detailed in https://hackmd.io/@jdahm/HJ-ZUn7nu, along with for loops. tldr; Cannot span more than one stage, which makes sense from a GridTools perspective. It's not a great fit for general gtscript, but the precedent is there because stencils are already limited by read-with-offset issues.

> Should nested loops be allowed? I made an example that fails in the parsing.

Difficult question. I would love to say "no". I think @leuty or @oelbert would be good people to ask. From my recollection, a single `while` loop is sufficient. Multiple nested `for` loops need to be supported while we do not have vector-based assignments.

I'd like to make sure we have a frontend error for all limitations before we merge. Let's make that a TODO (I cannot edit Eddie's initial message).

> Can you extend the parallel model with a section about while loops?

Another good TODO.

File: src/gtc/gtir.py [hunk 10]
```diff
@@ -268,3 +253,41 @@ def _cartesian_fieldaccess(node) -> bool:
 
 def _variablek_fieldaccess(node) -> bool:
     return isinstance(node, FieldAccess) and isinstance(node.offset, VariableKOffset)
+
+
+def _written_and_read_with_offset(
+    stmts: List[Stmt],
+) -> RootValidatorValuesType:
+    """
+    In the same VerticalLoop a field must not be written and read with a horizontal offset.
```
[the remark this answers is not in the document]
Comment (jdahm) [reply src/gtc/gtir.py:262]: Thanks for catching that. Fixed!

File: tests/test_unittest/test_gtc/gtir_utils.py [hunk 11]
```diff
@@ -37,7 +37,7 @@ class Meta:
 
     name = identifier(gtir.FieldAccess)
     offset = factory.SubFactory(CartesianOffsetFactory)
-    data_index: List[gtir.Expr] = []
+    data_index: List[gtir.Expr] = factory.List([])
```
[the remark this answers is not in the document]
Comment (jdahm) [reply tests/test_unittest/test_gtc/gtir_utils.py:40]: This is a relic of a previous iteration of the PR, where one test required this. I don't think we need it any more. If that's a factory it let's the user write `__0 = gtir.Expr(...)`

I will change that back, and we can check if the tests still pass.

File: tests/test_unittest/test_gtc/test_oir.py [hunk 12]
```diff
@@ -288,3 +289,7 @@ def test_assign_to_ik_fwd():
                 ],
             ),
         )
+
+
+def test_loop_mask():
+    StencilFactory(vertical_loops__0__sections__0__horizontal_executions__0__body__0=WhileFactory())
```
[the remark this answers is not in the document]
Comment (jdahm) [reply tests/test_unittest/test_gtc/test_oir.py:295]: Nothing useful 🙄  Except that it can instantiate one without error. That's definitely covered in another test.

File: tests/test_integration/test_code_generation.py [hunk 13]
```diff
@@ -383,7 +383,9 @@ def stencil_ijk(
             out_field = in_field[0, 0, 1] + in_field[0, 0, index_field + 1]
 
 
-@pytest.mark.parametrize("backend", OLD_BACKENDS)
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (jdahm) [reply tests/test_integration/test_code_generation.py:387]: Done!

File: src/gtc/gtir.py [hunk 14]
```diff
@@ -149,6 +145,23 @@ def verify_scalar_condition(cls, cond: Expr) -> Expr:
         return cond
 
 
+class While(common.While[Stmt, Expr], Stmt):
+    """
+    While loop with a field or scalar expression as condition.
+
+    No special rules apply.
```
[the remark this answers is not in the document]
Comment (jdahm) [reply src/gtc/gtir.py:152]: https://github.com/GridTools/concepts/pull/50

**Comment by leuty** [comment]
> I think @leuty or @oelbert would be good people to ask. From my recollection, a single while loop is sufficient. Multiple nested for loops need to be supported while we do not have vector-based assignments.

Not right now. We will encounter some rather nasty `for/while` nests later, and we'll have to see if we can re-write all of them properly.

**Comment by havogt** [comment]
Thanks, for linking again the hackmd document. Forgot about that. Let's discuss https://github.com/GridTools/concepts/pull/50 before I continue my review here.

File: src/gt4py/frontend/gtscript_frontend.py [hunk 15]
```diff
@@ -1270,15 +1270,20 @@ def visit_If(self, node: ast.If) -> list:
...
+        loc = gt_ir.Location.from_ast_node(node)
         if node.orelse:
             raise GTScriptSyntaxError("orelse is not supported on while loops")
         stmts = []
         for stmt in node.body:
             stmts.extend(self.visit(stmt))
+            if isinstance(stmts[-1], gt_ir.While):
+                raise GTScriptSyntaxError(
+                    "Nested while loops detected. This is unsupported.", loc=loc
+                )
```
Comment (havogt) [inline src/gt4py/frontend/gtscript_frontend.py:1282]: Please add a test, it's currently not working.

The fix is to either change line 1278 and check if `self.visit(stmt)` is a list or not or return a list from this method. I am not sure which is the right solution as I am not familiar with the frontend.

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gt4py/frontend/gtscript_frontend.py (hunk 15 above)
Comment (jdahm) [reply src/gt4py/frontend/gtscript_frontend.py]: After you mentioned that, I took another look. There's nothing blocking nested while loop, and I actually checked that it works correctly in gtcpp and the new gtc:numpy backend, so I removed the check here and cleaned it up.

**Comment by havogt** [comment]
bors try

File: src/gt4py/frontend/gtscript_frontend.py (hunk 15 above)
Comment (havogt) [reply src/gt4py/frontend/gtscript_frontend.py]: Looks good. The test that I had is now working.

**Review by havogt — CHANGES_REQUESTED** [review]
Please add a regression test for nested while loops. I think then it's good to go.

**Comment by jdahm** [comment]
Will do! Technically we should merge https://github.com/GridTools/gt4py/pull/628 first.

**Comment by jdahm** [comment]
bors try

**Review by havogt — APPROVED** [review]

**Comment by jdahm** [comment]
bors try

**Comment by jdahm** [comment]
bors try

**Comment by jdahm** [comment]
This PR implements the while loop construct in the GTC toolchain.
